### ISSUE TITLE
Find reasonable way to store Data

### ISSUE THREAD

**Issue by jennyahoffmann-zz** [jennyaberger/EasyLearning#5]
- [x] from Content

- [x] from Structure

**Comment by Malcolm-X** [comment]
**SQLITE**
The standard way to store structured data in android is [SQLite](https://sqlite.org/). SQLite is supported naturally ([see here](https://developer.android.com/training/basics/data-storage/databases.html)). Content from the db can be displayed easily in ListViews, GridViews and the likes using [CursorAdapter](https://developer.android.com/reference/android/widget/CursorAdapter.html) or by extending [BaseAdapter](https://developer.android.com/reference/android/widget/BaseAdapter.html).

As the name suggests, SQLite is a lightweight version of an SQL database. SQLite in most parts compatible to the SQL standard, the differences are summarized on [this page](https://sqlite.org/lang.html).
The most important difference are the missing FULL OUTER JOIN and RIGHT OUTER JOIN operations.

Besides that SQLite's underlying data structures differ, but all SQL-data-types are supported. E.g. creating a table row of type varchar(50) will result in a String field with a max length of 50 characters in SQL, while in SQLite this will be mapped to a simple TEXT field without any constraints.

**ORM / GREENDAO**
Although android has great out of the box support for database access through CursorAdapters, using an ObjectRelationalMapper can have some advantages when data should be manipulated in non standard ways (e.g. by synchronizing the data with an SQL-Server) or retrived with custom adapters.

ORM libraries will provide capabilities to persist and retrieve POJO's in the db. There are quite a few ORM libraries for Android: [ORMLite](http://ormlite.com/), [ActiveAndroid](http://www.activeandroid.com/), [GreenDAO](http://greenrobot.org/greendao/), etc.. I personally worked with GreenDAO both in a studend project and at work an had good overall experiences, but a huge drawback at that time was the incompatibility with the default android cursor api. I will elaborate the different current solutions in the following days. However, GreenDAO seems to be very popular and best performing ORM system in 2016 so far (at least according to their own benchmarks).

GreenDOA works by pregenerating classes for the database setup, access and classes that should be stored. These classes are generated in a seperate project and copied into the real apllication. This process can be integrated into the build-chain. The GreenDAO model is generated using Java which is straight forward.

Example code for GreenDAO can be found on the official website and in [this](https://github.com/Malcolm-X/buddycloudAndroid) repository (from a student project 4 years ago). If you cannot create a working .apk from the source, try downloading the provided .apk. In this project we implemented an Android client for the distributed social network buddycloud.org and use SQLite and GreenDOA to store and cache all data retrieved from the servers before displaying it. I build a small db framework that we can use as a template for our implementation. :

- A content item that can be stored / retrieved with GreenDAO [POJO](https://github.com/Malcolm-X/buddycloudAndroid/blob/master/buddycloud-android/src/tub/iosp/budcloand/framework/model/BCItem.java)  [ORM class](https://github.com/Malcolm-X/buddycloudAndroid/blob/master/buddycloud-android/src/tub/iosp/budcloand/framework/control/database/BCItemDao.java)

- A controlflow object that can be stored in the db [POJO](https://github.com/Malcolm-X/buddycloudAndroid/blob/master/buddycloud-android/src/tub/iosp/budcloand/framework/model/BCItem.java) [ORM class](https://github.com/Malcolm-X/buddycloudAndroid/blob/master/buddycloud-android/src/tub/iosp/budcloand/framework/control/database/CacheTimeFrameDao.java)

- db controller [Interface](https://github.com/Malcolm-X/buddycloudAndroid/blob/master/buddycloud-android/src/tub/iosp/budcloand/framework/control/database/BCDatabase.java) [ORM class](https://github.com/Malcolm-X/buddycloudAndroid/blob/master/buddycloud-android/src/tub/iosp/budcloand/framework/control/database/DefaultDatabase.java) [Singleton wrapper](https://github.com/Malcolm-X/buddycloudAndroid/blob/master/buddycloud-android/src/tub/iosp/budcloand/framework/control/database/DatabaseHelper.java)

- Use custome adapter and to display db content in ListView [VIEW(List)](https://github.com/Malcolm-X/buddycloudAndroid/blob/master/buddycloud-android/res/layout/activity_showpost.xml) [VIEW(ITEM)](https://github.com/Malcolm-X/buddycloudAndroid/blob/master/buddycloud-android/res/layout/activity_showpost_commentlist_item.xml) [ADAPTER](https://github.com/Malcolm-X/buddycloudAndroid/blob/master/buddycloud-android/src/tub/iosp/budcloand/framework/control/adapter/BCCommentListAdapter.java) [ORM class](https://github.com/Malcolm-X/buddycloudAndroid/blob/master/buddycloud-android/src/tub/iosp/budcloand/framework/control/database/DefaultDatabase.java) [Very long Activity](https://github.com/Malcolm-X/buddycloudAndroid/blob/master/buddycloud-android/src/tub/iosp/budcloand/activities/BasicActivity.java)

**FILE STORAGE**
Unstructured data, binaries and view artifacts (e.g. html, css,...) can be stored on the local storage. Storing structured data in files (e.g. XML) is possible but not preferred.

**CONCLUSION**
TL/DR:
I suggest using SQLite with an ORM library, maybe GreenDAO.

**Comment by Malcolm-X** [comment]
We will use GreenDao
